compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs

```diff
@@ -429,7 +350,49 @@ impl<'a> ModCollector<'a> {
                 errors.push((error.into(), self.file_id));
             }
 
-            errors.append(&mut self.check_for_duplicate_trait_items(&trait_definition.items));
+            for trait_item in &trait_definition.items {
+                match trait_item {
+                    TraitItem::Function { name, .. } => {
+                        if let Err((first_def, second_def)) = self.def_collector.def_map.modules
+                            [id.0.local_id.0]
+                            .declare_function(name.clone(), FuncId::dummy_id())
```

Comment:
It looks like we create the actual function ids later on line 411. Why not do this here so we don't have to use dummy ids? In general, I'd like to avoid dummy ids whenever possible.

Also, do we really need a separate loop here? Can we merge this with the one below?

Reply:
For a FuncId we need a a concrete function with a body, which we don't have at this stage (unless we have a default implementation but even then, using it is wrong since `MyTrait::method(foo, bar)` would always resolve to it, rather than pick the correct impl).

Ultimately to support `MyTrait::method(foo, bar)` I believe we'd want `method` to resolve to a `TraitMethodId` (trait id + method index), which *should* give us the info needed to pick the correct impl later on, but I'm not 100% sure on that, so I told @nickysn it's probably best to use a dummy and we'll figure it out when we do the `Triat::method() syntax` - for now this is only used to throw the duplicate error, so it doesn't really matter what the value is.

Reply:
Yeah, but I guess it doesn't hurt to just always call def_interner.push_empty_fn() to get an id, even if the function doesn't have a body. We'll just skip adding an unresolved function after that, if it doesn't have a body. I'll try this and see if it breaks anything...

Reply:
Ok, I merged the loops, refactored the code a little bit and used push_empty_fn() and push_empty_global() to get non-dummy ids for the trait functions and consts. However, there's no equivalent push_empty_type_alias(), so I haven't done this for types. I'd rather have this implemented by @alexvitkov, because he has started work on associated types in a branch. I don't want to refactor the type alias and related structs struct to support empty types, because this might cause conflicts with his work. I've just added a TODO comment, so this is not forgotten. Let's merge this now, since associated types are a post-MVP feature, and they're not working, yet, so pushing dummy ids for them can't break what's not working anyway :) However, merging this pull request is important for the traits MVP, because trait functions are an MVP feature and they need a duplicate check. Also because, it includes another unrelated bug fix for traits in different crates.